### PR TITLE
Form functions

### PR DESCRIPTION
**Pull request by adammichaelwood**
#### What is included in this PR?

 - Function and operator reference
 - Reusable form examples pulled into includes
 - conf.py updated to use JS domain by default (See below)
 - custom css to make function reference easier to read
 - fixed a few form examples (form-logic and form-repeats) that were either wrong or had typos

#### What issues need to be opened?

Opening an ongoing issue to add useful examples.

##### Minor - Hard to follow existing references

A few of the functions listed in the XForm spec weren't clear, and a few weren't relevant --- tracking that down was a minor pain.

cf: https://opendatakit.github.io/xforms-spec/#xpath-functions

##### Sort of major - Domain work

In Sphinx, function references like this are done using language specific domains.
 cf: http://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html

There is no XForm/XPath domain. I tried using the default domain (Python), but the syntax didn't work out. (Python doesn't allow hyphens, and does namespacing differently.) Then I tried to write my own domain, using the Python domain as a starting point. This was a terrible mistake.

Finally, I discovered that the JS domain was *good enough* for current purposes. So `conf.py` has been updated to use that as the default. It might become a problem in the future, depending on how much more XForm/Xpath stuff makes it into the docs. Also, it may become an issue if we decide to pull the ODK XForm spec into Sphinx.

##### Sidenote problem -- git rebase

I did the following:

 - rebased all my work on top of a fresh pull of master
 - fixed conflicts
 - commited locally
 - pushed to GH (--force)

And I *still* had to come fix a merge conflict (The same one) here in GH. There's something in this rebase process I'm missing. It wasn't a big deal, but it's a little annoying.

### CODE REVIEW

**Comment by yanokwa** [comment]
This is a great start. I think it's not quite glanceable. I'd like to see a quick table up front sorted by name or organized by theme. It reads a little technical to me and so maybe an example for each might help? https://opendatakit.org/help/form-design/binding I think was pretty good at that.

**Comment by adammichaelwood** [comment]
Still working on examples. 
But --- does this organization make sense?
Section naming here is weird and hard, so suggestion welcome.

**Comment by adammichaelwood** [comment]
Oh, also...

I think this page should also have operators on it.
If so, does the main title change?

File: odk1-src/form-functions.rst [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+
+
+.. function:: regex(string, expression)
+
+  Returns ``True`` if ``string`` is an exact match for ``expression``.
+
+  .. seealso:: :doc:`form-regex`
+
+
+.. function:: contains(haystack, needle)
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: My guess is that this is a very European figure of speech. Maybe string/substring or string/keyword? What do other docs do?

File: odk1-src/form-functions.rst [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+  Returns ``then`` if ``expression`` evaluates to ``True``. 
+  Otherwise, returns ``else``.
+
+  
+.. function:: not(arg)
+
+  Returns the opposite of ``boolean(arg)``.
+
+  
+.. function:: coalesce(arg, arg)
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: This feels more like combining strings.

File: odk1-src/form-functions.rst [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  
+  
+  
+.. function:: format-date(date, format)
+
+  Returns ``date`` as a string formatted as defined by ``format``.
+  
+  The following identifiers are used in the ``format`` string:
+  
+  ``%Y``
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: Can we put this stuff in a table? Current layout feels wasteful.

File: odk1-src/form-functions.rst [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+.. function:: position(xpath)
+
+  Returns an integer equal to the 1-indexed position of the current node
+  within the node defined by ``xpath``.
+  
+  Most often this is used in the form ``position(..)``
+  to identify the current iteration index.  
+  
+  
+String handling
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: Might be OK to just call this string. Since you don't call it "math and numbers handling"

File: odk1-src/form-functions.rst [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+.. function:: exp10(number arg)
+
+  Returns ``10^x``.
+
+.. function:: pi()
+
+  Returns an approximation of the mathematical constant π.
+
+
+Control flow
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: This feels out of place here. Maybe it should go first or second (after form data)

File: odk1-src/form-functions.rst [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+  Returns the response value of question ``name``
+  from the repeat-group ``group``,
+  in iteration ``i``.
+  
+  Nested repeat groups can be accessed 
+  using the ``sub`` and ``sub_sub`` parameters,
+  which can be extended to three levels of nesting.
+
+  
+Counting responses
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: Feels mathy to me.

File: odk1-src/form-functions.rst [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+.. function:: count(nodeset)
+
+  Returns the number of items in ``nodeset``. This can be used to count the number of repetitions in a :ref:`repeat group <repeats>`.
+
+.. function:: count-non-empty(nodeset)
+
+  Returns the number of non-empty members of ``nodeset``.
+
+ 
+Doing math with responses
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: Feels mathy to me. It even says math in the title :)

File: odk1-src/form-functions.rst [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+Accessing information about the form
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
+ 
+.. function:: jr:choice-name(select_question, choice_name)
+
+  Returns the label value, in the active language, associated with the ``choice_name`` in the list of choices for the ``select_question``.
+
+.. function:: jr:itext(name) 
+
+  Returns the itext in the active language for the named item.
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: The itext here seems self-referential. Maybe just text?

File: odk1-src/form-functions.rst [hunk 9]
```diff
@@ -0,0 +1,502 @@
...
+
+  Returns the sum of the members of ``nodeset``.
+
+.. function:: max(nodeset)
+
+  Returns the largest member of ``nodeset``.
+
+.. function:: min(nodeset)
+
+  Returns the smallest memberd of ``nodeset``.
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: members

File: odk1-src/form-functions.rst [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+
+.. contents::
+  :local:   
+
+
+  
+Form data 
+----------------
+
+Accessing previous responses
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: Feels like most of this has to do with selects, so why not have a selects section, then have just a repeats section (even if it has one item). Then could put the mathy bits in the math section under their own subsection, move position and itext to utility.

File: odk1-src/form-functions.rst [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+.. function:: jr:itext(name) 
+
+  Returns the itext in the active language for the named item.
+  
+.. function:: position(xpath)
+
+  Returns an integer equal to the 1-indexed position of the current node
+  within the node defined by ``xpath``.
+  
+  Most often this is used in the form ``position(..)``
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: I love this. Helps the user understand the common case.

File: odk1-src/form-functions.rst [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+    select_multiple cancer_types, what_cancer, What type of cancer have you experienced?, Select all that apply.,"selected(${what_issues}, 'cancer')",
+  
+  .. seealso:: :ref:`The complete example of this form, with a video. <complex-conditional-example>`
+  
+  
+.. function:: selected-at(multi_select_question, n)
+
+  Returns the :th:`name` of the :arg:`n`\th selected choice of the :arg:`multi_select_question`. (Selected choices are zero-indexed.)
+  
+  .. note::
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: Good!

File: odk1-src/form-functions.rst [hunk 13]
```diff
@@ -0,0 +1,502 @@
+Form Functions
+================
+
+:ref:`expressions` in :ref:`calculations`, :ref:`constraints <constraints>`, and :ref:`relevants <relevants>`
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: Can we make calculations lower case in the rendered view? Looks out of place.

**Comment by adammichaelwood** [comment]
> missing checklist(), weighted-checklist(), and property()

Yes, I couldn't figure out what those did when I was reading the XForm spec, 
but now I'm looking at the old ODK bindings page you linked,
and it is more clear there.

File: odk1-src/form-functions.rst (hunk 6 above)
Comment (adammichaelwood) [reply odk1-src/form-functions.rst]: yeah, i wondered about that...

File: odk1-src/form-functions.rst (hunk 3 above)
Comment (adammichaelwood) [reply odk1-src/form-functions.rst]: i like the semantics of definition lists, so maybe a compact view...

File: odk1-src/form-functions.rst (hunk 1 above)
Comment (adammichaelwood) [reply odk1-src/form-functions.rst]: String/substring seems like a good alternative.

File: odk1-src/form-functions.rst (hunk 2 above)
Comment (adammichaelwood) [reply odk1-src/form-functions.rst]: It would work on non-string values, though too --- right?

**Comment by adammichaelwood** [comment]
see if that organizations structure makes more sense.

**Comment by adammichaelwood** [comment]
This is (I hope) ready to go.
@yanokwa

File: odk1-src/form-functions.rst [hunk 14]
```diff
@@ -0,0 +1,674 @@
...
+---------------
+
+.. csv-table::
+  :header: , Explanation, Example, Notes
+  
+  \+, addition, ${salary_income} + ${self_employed_income}, Numbers only; does not concatenate strings.
+  \-, subtraction, ${income} - ${expenses},
+  \*, multiplication, ${bill} * 1.18,
+  div, division, ${percent_int} div 100, 
+  mod, `modulo`_ (division remainder), (. mod 2) = 0,
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: The use of `.` here sticks out.  I think something like `(15 mod 2) = 1` would be clearer.

File: odk1-src/form-functions.rst [hunk 15]
```diff
@@ -0,0 +1,674 @@
...
+
+  Returns ``True`` if 
+  the sum of the :arg:`weight`\ s 
+  of each :arg:`response` that is exactly the string "yes"
+  is between :arg:`min` and :arg:`max`, inclusive.
+  
+  Set :arg:`min` or :arg:`max` to ``-1`` to make the argument not 
+    
+
+.. _xform-operators:
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: I really like this section. Nice and clear. But it's the most common thing people would look for, so I think it should go first.

And now that I've written that, maybe it should just be in its own section after Form Logic

File: odk1-src/form-functions.rst [hunk 16]
```diff
@@ -0,0 +1,674 @@
...
+  Returns :arg:`then` if :arg:`expression` evaluates to ``True``. 
+  Otherwise, returns :arg:`else`.
+
+  
+.. function:: not(arg)
+
+  Returns the opposite of :func:`boolean(arg) <boolean>`.
+
+  
+.. function:: coalesce(arg, arg)
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: coalesce and boolean feel to my like utility functions. Or rather, they feel like functions that no one uses and they are front and center.

File: odk1-src/form-functions.rst [hunk 17]
```diff
@@ -0,0 +1,674 @@
...
+
+.. function:: number(string)
+
+  Converts a :arg:`string` of digits into a number value.
+
+.. function:: string(arg)
+
+   Converts :arg:`arg` to a string.
+
+.. _math-functions:
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: I was expecting to see count/max/min. Can we put a note here saying that there are some math operations that only work on nodesets?

File: odk1-src/form-functions.rst [hunk 18]
```diff
@@ -0,0 +1,674 @@
...
+  Returns an integer equal to the 1-indexed position of the current node
+  within the node defined by :arg:`xpath`.
+  
+  Most often this is used in the form :tc:`position(..)`
+  to identify the current iteration index
+  within a repeat group.  
+  
+  .. include:: incl/form-examples/parallel-repeat-groups.rst
+
+.. function:: uuid([length])
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: I'd put this above the checklist functions since it's not that used.

File: odk1-src/form-functions.rst [hunk 19]
```diff
@@ -0,0 +1,674 @@
...
+  nodeset
+  prefs
+  seleccionados
+  te
+  tres
+  π
+  
+  
+	
+******************************
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: I doubt this would be possible, but it'd be nice to see an alphabetical list of all the things in a table to make it glanceable.

Maybe a link to https://opendatakit.github.io/xforms-spec/#xpath-functions would be good enough?

File: odk1-src/form-functions.rst [hunk 20]
```diff
@@ -0,0 +1,674 @@
...
+  .. note::
+  
+    This function returns the :th:`name`, not the :th:`label`,
+    of the selected choice.
+    To get the label in the current language,
+    use :func:`jr:choice-name`.
+  
+  Returns an empty string if the index does not exist.
+  
+  .. image:: /img/form-functions/selected-at-0.*
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: Can all these be available as a detail? It makes it hard to skim.

File: odk1-src/form-functions.rst [hunk 21]
```diff
@@ -0,0 +1,674 @@
...
+  tres
+  π
+  
+  
+	
+******************************
+Form Functions and Operators
+******************************
+
+:ref:`expressions` in :ref:`calculations`, :ref:`constraints <constraints>`, and :ref:`relevants <relevants>`
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: Possible to make this lowercase in the display?

**Comment by adammichaelwood** [comment]
Addressed all of the above (I think.)

I can pull operators into their own doc if you want. For now, I just moved them to the top of the page.

**Comment by adammichaelwood** [comment]
(Oh -- except the details thing. I haven't done that. Will add if you want me to. I don't mind scrolling as much as you do, I think.)

**Comment by adammichaelwood** [comment]
ok - i changed my mind.
I tried the details thing and it is totally better.

File: odk1-src/form-functions.rst [hunk 22]
```diff
@@ -0,0 +1,719 @@
...
+  
+	
+******************************
+Form Functions and Operators
+******************************
+
+:ref:`expressions` in :ref:`calculations <calculations>`, :ref:`constraints <constraints>`, and :ref:`relevants <relevants>`
+can contain operators and functions.
+
+.. seealso:: `XPath Functions <https://opendatakit.github.io/xforms-spec/#xpath-functions>`_
```
Comment (lognaturel) [inline odk1-src/form-functions.rst]: Maybe "XPath functions in the ODK XForms specification" to make it clear what that doc is?

File: odk1-src/form-functions.rst [hunk 23]
```diff
@@ -0,0 +1,719 @@
...
+Boolean operators
+------------------
+
+.. csv-table::
+  :header: , Explanation, Example
+
+  and, ``True`` if the expressions before and after are ``True``, ${age} > -1 and ${age} < 120
+  or, ``True`` if either of the expressions before or after are ``True``, ${age} < 19 or ${age} > 64
+  
+.. _comparison-operators:
```
Comment (lognaturel) [inline odk1-src/form-functions.rst]: I would tend to put comparison operators before boolean operators. Most people aren't familiar with the word "boolean". Perhaps there can be a short sentence when introducing comparison operators that says that the result is always `true()` or `false()` (boolean). Then the boolean operators introduction can say they join two boolean expressions (that evaluate to `true()` or `false()`).

File: odk1-src/form-functions.rst [hunk 24]
```diff
@@ -0,0 +1,719 @@
...
+  =, equal to, ${enrolled} = 'yes', Can compare numbers or strings.
+  !=, not equal to, ${enrolled} != 'yes', Can compare numbers or strings.
+  >, greater than, ${age} > 17, 
+  >=, greater than or equal to, ${age} >= 18,
+  <, less than, ${age} < 65, 
+  <=, less than or equal to, ${age} <= 64,
+
+.. _math-operators:
+  
+Math operators
```
Comment (lognaturel) [inline odk1-src/form-functions.rst]: I would put math operators first since they are familiar.

File: odk1-src/form-functions.rst [hunk 25]
```diff
@@ -0,0 +1,719 @@
...
+
+  
+.. _xform-operators:
+
+Operators
+==========
+
+.. _utility-operators:
+
+Utility operators
```
Comment (lognaturel) [inline odk1-src/form-functions.rst]: These are XPath references to the current and containing node. They can be used to build up relative paths (e.g. ../../food). It's ok to put them in this section but I think they should have a little more context. See https://opendatakit.github.io/xforms-spec/#xpath-paths. Perhaps this should go last since it's a little more esoteric.

File: odk1-src/form-functions.rst [hunk 26]
```diff
@@ -0,0 +1,719 @@
...
+  questions inside :ref:`repeat groups <repeat-functions>`.
+
+.. _select-functions:
+  
+Select questions
+~~~~~~~~~~~~~~~~~~~
+
+.. function:: selected(select_question, choice_name)
+
+  Returns ``True`` if :arg:`choice_name`
```
Comment (lognaturel) [inline odk1-src/form-functions.rst]: I think it's useful to describe this a little more generally and mention that it searches for the second parameter in the space-separated list specified by the first parameter. There are all kinds of creative ways to use `selected` and `selected-at` with that more general concept.

File: odk1-src/form-functions.rst [hunk 27]
```diff
@@ -0,0 +1,719 @@
...
+~~~~~~~~~~~~~~~~
+
+.. admonition:: Helpful terms
+
+  .. glossary::
+    :sorted:
+
+    nodeset
+
+      A collection of response values.
```
Comment (lognaturel) [inline odk1-src/form-functions.rst]: It may be worth mentioning that this is as in XPath -- a set of any XML nodes. You can build a nodeset a number of ways. For example, if you have nodes `/data/name` and `/data/age`, you can do `/data/name | /data/age` to build up a nodeset that contains both nodes. Functions like `randomize` return nodesets not necessarily of responses (generally of choices).

File: odk1-src/form-functions.rst [hunk 28]
```diff
@@ -0,0 +1,719 @@
...
+
+.. function:: boolean-from-string(string)
+
+  Returns ``True`` if :arg:`string` is "true" or "1".
+  Otherwise, ``False``.
+
+
+.. function:: number(string)
+
+  Converts a :arg:`string` of digits into a number value.
```
Comment (lognaturel) [inline odk1-src/form-functions.rst]: Might want to say the value is `NaN` if it can't be converted to a number. 1 if `true()`, 0 if `false()`. https://www.w3.org/TR/1999/REC-xpath-19991116/#function-number

File: odk1-src/form-functions.rst [hunk 29]
```diff
@@ -0,0 +1,719 @@
...
+Number handling
+~~~~~~~~~~~~~~~~~
+
+.. function:: round(number, places)
+
+  Rounds a decimal :arg:`number` to some number of decimal :arg:`places`.
+
+.. function:: int(arg) 	
+
+  Converts :arg:`arg` to an integer.
```
Comment (lognaturel) [inline odk1-src/form-functions.rst]: To be explicit, I might say "truncates the fractional part of a decimal value to make it an integer"

File: odk1-src/form-functions.rst [hunk 30]
```diff
@@ -0,0 +1,719 @@
...
+Date and time
+----------------
+
+.. function:: today()
+
+  Returns the current date without a time component.
+
+.. function:: now()
+
+  Returns the current datetime in the current time zone.
```
Comment (lognaturel) [inline odk1-src/form-functions.rst]: in iso 8601 format including the timezone - https://en.wikipedia.org/wiki/ISO_8601

File: odk1-src/form-functions.rst [hunk 31]
```diff
@@ -0,0 +1,719 @@
...
+.. function:: random()
+
+  Returns a random number between 0.0 (inclusive) and 1.0 (exclusive).
+
+  
+.. function:: randomize(nodeset[, seed]) 	
+
+  Returns a shuffled :arg:`nodeset`.
+  
+  A shuffle with a numeric :arg:`seed` is deterministic and reproducible.
```
Comment (lognaturel) [inline odk1-src/form-functions.rst]: May want to suggest saving a seed in a previous calculate (e.g. using `random` or `decimal-date-time(now())`) and using it so that the order can be recreated.

File: odk1-src/form-logic.rst [hunk 32]
```diff
@@ -316,7 +308,7 @@ Often, comparison `operators`_ are used in relevance expressions. For example:
 :tc:`${has_children} = 'yes'`
   True if the answer to :tc:`has_children` was ``yes``.
   
-Relevance expressions can also use `XPath functions`_.
+Relevance expressions can also use :doc:`XPath functions <form-functions>`.
```
Comment (lognaturel) [inline odk1-src/form-logic.rst]: May be best to use just `functions` here. There's a mix of XPath and XForms and madeup functions, really.

File: odk1-src/form-functions.rst [hunk 33]
```diff
@@ -0,0 +1,719 @@
...
+  
+.. function:: checklist(min, max, response[, response[, response [, ... ]]])
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: I'd move checklist/weighted-checklist to the bottom, above true/false because they are so infrequently used.

File: odk1-src/form-functions.rst [hunk 34]
```diff
@@ -0,0 +1,719 @@
...
+  prefs
+  seleccionados
+  te
+  tres
+  π
+  
+  
+	
+******************************
+Form Functions and Operators
```
Comment (yanokwa) [inline odk1-src/form-functions.rst]: Let's split into two documents.

Comment (lognaturel) [reply odk1-src/form-functions.rst]: I'm not so sure about this one! Operators can join the results of functions. I think of this document as being about the building blocks of expressions.

Comment (yanokwa) [reply odk1-src/form-functions.rst]: If y'all don't want to split, that's fine by me. But at the minimum maybe it should be called `Form Operators and Functions` to match the order?

Comment (adammichaelwood) [reply odk1-src/form-functions.rst]: I agree with this last point. I'll change the file name too.

File: odk1-src/form-functions.rst (hunk 27 above)
Comment (adammichaelwood) [reply odk1-src/form-functions.rst]: Could one do:

``${question_name_1} | ${question_name_2}``

?

Comment (adammichaelwood) [reply odk1-src/form-functions.rst]: > Functions like randomize return nodesets not necessarily of responses (generally of choices).

File: odk1-src/form-functions.rst (hunk 25 above)
Comment (adammichaelwood) [reply odk1-src/form-functions.rst]: > Perhaps this should go last since it's a little more esoteric

I don't know... in their simple "naive" sense, they get used all the time. You can't really deal with constrains without `.`, and `position(..)` is pretty common inside repeats.

I think we need a doc or section on XPath, separate from this. Once that is in place, I think linking from here to there with a note ("These are actually XPath references.... etc...") would be a good idea. For now, maybe a similar note that links to https://opendatakit.github.io/xforms-spec/#xpath-paths.

Comment (adammichaelwood) [reply odk1-src/form-functions.rst]: I'm adding a note, and --- after trying a few arrangements --- ended up putting them last as suggested.

**Comment by lognaturel** [comment]
Looks good to me!
